fdbserver/DataDistribution.actor.cpp

```diff
@@ -1512,219 +1631,690 @@ ACTOR Future<Void> ddGetMetrics(GetDataDistributorMetricsRequest req,
-			    .detail("AuditType", auditState.type);
-			if (e.code() == error_code_audit_storage_error) {
-				auditState.setPhase(AuditPhase::Error);
-				wait(persistAuditState(self->txnProcessor->context(), auditState));
-			} else if (retryTime > 10) {
-				auditState.setPhase(AuditPhase::Failed);
-				wait(persistAuditState(self->txnProcessor->context(), auditState));
-			} else if (e.code() != error_code_actor_cancelled) {
-				wait(delay(30));
-				retryTime++;
+			TraceEvent(SevVerbose, "WaitForAuditStorage", self->ddId)
+			    .detail("AuditID", auditID)
+			    .detail("AuditType", auditType);
+			// auditMap keeps following invariants:
+			// (1) Any alive audit storage must be in auditMap
+			// (2) Any audit of auditMap must be alive
+			if (auditExistInAuditMap(self, auditType, auditID)) {
+				wait(delay(1));
 				continue;
+			} else {
+				TraceEvent(SevInfo, "WaitForAuditStorage", self->ddId)
+				    .detail("AuditID", auditID)
+				    .detail("AuditType", auditType);
+				break;
+			}
+		} catch (Error& e) {
+			if (e.code() == error_code_actor_cancelled) {
+				throw e;
 			}
+			TraceEvent(SevDebug, "WaitForAuditStorage", self->ddId)
+			    .errorUnsuppressed(e)
+			    .detail("AuditID", auditID)
+			    .detail("AuditType", auditType);
+			continue;
 		}
-		break;
 	}
-	self->audits[auditState.getType()].erase(audit->id);
 	return Void();
 }
 
-ACTOR Future<Void> auditStorage(Reference<DataDistributor> self, TriggerAuditRequest req) {
-	std::vector<Future<Void>> fs;
-	fs.push_back(self->auditInitialized.getFuture());
-	fs.push_back(self->initialized.getFuture());
-	wait(waitForAll(fs));
+// runAuditStorage is the only entry to start an Audit entity
+// Three scenarios when using runAuditStorage:
+// (1) When DD receives an Audit request;
+// (2) When DD restarts and resume an Audit;
+// (3) When an Audit gets failed and retries.
+// runAuditStorage is a non-flow function which starts an audit for auditState
+// with three steps (the three steps are atomic):
+// (1) Create audit data structure, (2) register it to dd->audits, (3) run auditStorageCore
+void runAuditStorage(Reference<DataDistributor> self,
+                     AuditStorageState auditState,
+                     int retryCount,
+                     std::string context) {
+	if (auditState.getType() != AuditType::ValidateHA && auditState.getType() != AuditType::ValidateReplica &&
+	    auditState.getType() != AuditType::ValidateLocationMetadata &&
+	    auditState.getType() != AuditType::ValidateStorageServerShard) {
+		throw not_implemented();
+	}
+	ASSERT(auditState.id.isValid());
+	std::shared_ptr<DDAudit> audit = std::make_shared<DDAudit>(auditState.id, auditState.range, auditState.getType());
+	if (self->auditInitialized.getFuture().isReady()) {
+		ASSERT(audit->coreState.getPhase() == AuditPhase::Invalid);
+	}
+	// for invalid phase, make it running, used for resume audit when dd restarts
+	if (audit->coreState.getPhase() == AuditPhase::Invalid) {
```

Comment:
Same here, we are not reusing `auditState`

Reply:
I think the code would be more explicit to not reuse a `state`. I think reusing the state would introduce some implication for future changes of (such as) the state definition.

Reply:
Yes. Thanks!